### PR TITLE
feat(react): pass props and child render props

### PR DESCRIPTION
**Pull request by zgover**
# Description

- fix: respect of children prop to support subscribing to contexts through child components rather than only `FormTemplate` or fields.
- feat: optimized portability of `FormRender` by supporting `FormTemplate` pass prop through `FormTemplateProps`.
- feat: overridable form rendering by passing children as a render function.

Please include a summary of the change.

**Schema** *(if applicable)*

```jsx
// Children prop as ReactElement
const MyOtherChild = ({allowReset}) => {
    const {onReset} = useFormApi()
   return !allowReset ? null : (
       <button onClick={() => onReset()}/>Reset</button>
   )
}
<FormRenderer
    FormTemplate={AuthFormTemplateComponent}
    componentMapper={simpleComponentMapper}
    onSubmit={handleFormSubmit}
    schema={formSchema}
    subscription={{values: true}}
    clearOnUnmount
>
    <MyOtherChild allowReset/>
</FormRenderer>
```

```jsx
// Template pass props
<FormRenderer
    FormTemplate={AuthFormTemplateComponent}
    FormTemplateProps={{type: 'signup'}} // signout | signup => form variant
    componentMapper={simpleComponentMapper}
    onSubmit={handleFormSubmit}
    schema={formSchema}
    subscription={{values: true}}
    clearOnUnmount
  />
```

```jsx
// Children render prop
<FormRenderer
    componentMapper={simpleComponentMapper}
    onSubmit={handleFormSubmit}
    schema={formSchema}
    subscription={{values: true}}
    clearOnUnmount
>
    {({formFields, schema}) => (
        <AuthFormTemplateComponent
            schema={schema}
            formFields={formFields}
            type={'signup'}
       />
    )}
</FormRenderer>
```

**Checklist:** *(please see [documentation page](https://data-driven-forms.org/development-setup) for more information)*

- [ ] `Yarn build` passes
- [ ] `Yarn lint` passes
- [ ] `Yarn test` passes
- [ ] Test coverage for new code *(if applicable)*
- [ ] Documentation update *(if applicable)*
- [ ] Correct commit message
   - format `fix|feat({scope}): {description}`
   - i.e. `fix(pf3): wizard correctly handles next button`
   - fix will release a new \_.\_.X version
   - feat will release a new \_.X.\_ version (use when you introduce new features)
     - we want to avoid any breaking changes, please contact us, if there is no way how to avoid them
   - scope: package
   - if you update the documentation or tests, do not use this format
     - i.e. `Fix button on documenation example page`

### CODE REVIEW

**Comment by zgover** [comment]
@Hyperkid123 the linter is way too strict on whitespace, it does not want to allow a multiline ternary boolean. Personally, I find it very disorganized whenever a lines span 100 while using xml syntax. How do you want to proceed with this?

**Review by Hyperkid123 — COMMENTED** [review]
@zgover thanks for the PR. I am all for this feature however we need a few changes.

1. By removing the required prop from the form template, we now have to check if the `FormTemplate` or `children` prop exists. You can probably add a runtime check as we do with the schema validator on the top of the component.
2. Add a unit test
3. We have to add documentation for this change
4. The linter issue is not about maximum line length but because the "else" branch of the ternary operator is a multiline expression (JSX). I think the proposed fix is not that bad.

**Comment by zgover** [comment]
@Hyperkid123 before I proceed—see last commit; I want to make sure you're ok with these changes.

**Comment by Hyperkid123** [comment]
@zgover I am more than OK with the last commit. these changes are something we should have done a long time ago.

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (rvsia) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:89]: `setRegisteredFields` cannot be changed, so empty array is enough here

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (rvsia) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:53]: personally I would return `undefined` instead of `void 0`, we do not use it anywhere else in the codebase

Comment (rvsia) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:53]: personally I would return `undefined` instead of `void 0`, we do not use it anywhere else in the codebase

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (rvsia) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:95]: ^^^

Comment (rvsia) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:95]: ^^^

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (rvsia) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:42]: Not sure if dependency array will work well with an object, the same applies for mutators/decorators (arrays)

Comment (rvsia) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:42]: Not sure if dependency array will work well with an object, the same applies for mutators/decorators (arrays)

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (zgover) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js]: Updated in last commit

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (zgover) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js]: Updated in last commit

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (zgover) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js]: Updated in last commit

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (zgover) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js:42]: Elaborate?

**Comment by zgover** [comment]
Last commit uncluded @rvsia suggestions, haven't had a chance to really look over the build error logs yet but just at the short glance looks like it may just be tests that need to be updated or a fix is due in one of my updates.

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (rvsia) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js:42]: The `useMemo` will run comparison between two arrays/objects. In most cases these arrays/objects will be recreated as a unnamed prop on every render, so they will be different even nothing changes. But it's not a big deal.

**Comment by zgover** [comment]
@Hyperkid123 @rvsia, I'm not too familiar with the current tests that are set up and entire submit flow. So it may be easier for y'all to determine the major issues here in the failed build.

For example it appears the submit handler is being called but there seems to be an additional field value of `undefined` in the values being passed to all of them:
```
 FAIL  packages/suir-component-mapper/src/tests/wizard.test.js 
  ● wizard › conditional submit

    expect(jest.fn()).toHaveBeenCalledWith(...expected)

    - Expected
    + Received

      Object {
    -   "aws": "aws",
    -   "summary": "summary",
    +   "undefined": undefined,
      },

...
```

Secondly, it appears the only validation for the schema error render is checking the console logs:
```
FAIL  packages/react-form-renderer/src/tests/form-renderer/form-renderer.test.js
  ● <FormRenderer /> › should render errorComponent form from schema

    expect(jest.fn()).toHaveBeenCalledWith(...expected)

    Expected: "error: ", Any<String>

    Number of calls: 0

      105 |     expect(screen.getByText('Form could not be rendered, because of invalid form schema.')).toBeInTheDocument();
      106 |     expect(spy).toHaveBeenCalled();
    > 107 |     expect(logSpy).toHaveBeenCalledWith('error: ', expect.any(String));
          |                    ^
      108 |
      109 |     console = _console; // eslint-disable-line
      110 |   });

      at Object.<anonymous> (packages/react-form-renderer/src/tests/form-renderer/form-renderer.test.js:107:20)
```

**Comment by Hyperkid123** [comment]
@zgover 

> For example it appears the submit handler is being called but there seems to be an additional field value of undefined in the values being passed to all of them:

```
 FAIL  packages/suir-component-mapper/src/tests/wizard.test.js 
  ● wizard › conditional submit
```

So in this test, the submit callback does not receive correct values. Not only there is an extra `undefined` but the values that should have been in the values object are entirely missing. So the submit call back is broken

I will checkout the PR and see why the submit function has invalid argument

> Secondly, it appears the only validation for the schema error render is checking the console logs:

That is not the case. The expect a couple of lines above also checks if the error component was rendered.
```
expect(screen.getByText('Form could not be rendered, because of invalid form schema.')).toBeInTheDocument();
```

**Comment by Hyperkid123** [comment]
@zgover the issue with the wizard test is in the `internalGetRegisteredFields` function. Let me check the changes and why is not doing what it's supposed to.

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (Hyperkid123) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:99]: @zgover  This function is missing `return` that is why the submit tests in wizard are failing.

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (Hyperkid123) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:67]: The original code is calling the `console.log` with just the error message without the whole stack
```js
  try {
    const validatorTypes = Object.keys(validatorMapperMerged);
    const actionTypes = actionMapper ? Object.keys(actionMapper) : [];
    defaultSchemaValidator(schema, componentMapper, validatorTypes, actionTypes, schemaValidatorMapper);
  } catch (error) {
    schemaError = error;
    // eslint-disable-next-line no-console
    console.error(error);
    // eslint-disable-next-line no-console
    console.log('error: ', error.message);
  }
  ```
That is why the second test is falling. I am honestly OK with just using the `console.error`. @rvsia do you have any objections to removing the `console.log` part? If no we can just drop the expect from the test

Comment (zgover) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js:67]: @rvsia ?

Comment (Hyperkid123) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js:67]: @zgover sorry I was on vacation. Just delete the console log check from tests. It's not necessary to log the same message multiple times.

**Comment by zgover** [comment]
@Hyperkid123 Looks like the remaining build errors are caused from outdated libs and implementation as I don't recall these being in here before.

```
console.warn
    Warning: The Filterable field has been deprecated on the MultiSelect object. Please import and use FilterableMultiSelect directly.
```

```
DualListSelect › sort value
```

```
DualListSelect › sort options
```

```
console.error
    Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
        at Popover (/home/circleci/react-forms/node_modules/@patternfly/react-core/src/components/Popover/Popover.tsx:216:3)
```

```
console.error
    Warning: React does not recognize the `defaultTime` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `defaulttime` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
        at div
        at div
        at TimePicker (/home/circleci/react-forms/node_modules/@patternfly/react-core/src/components/TimePicker/TimePicker.tsx:88:5)
```

```
console.error
    Warning: Unknown event handler property `onValueChange`. It will be ignored.
        at div
        at Slider (/home/circleci/react-forms/node_modules/@patternfly/react-core/src/components/Slider/Slider.tsx:66:3)
```

```
console.error
    Warning: React does not recognize the `currentValue` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `currentvalue` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
        at div
        at Slider (/home/circleci/react-forms/node_modules/@patternfly/react-core/src/components/Slider/Slider.tsx:66:3)
```

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (Hyperkid123) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:56]: @zgover This causes the issue with the unexpected `Cancel` button in FormTemplates.

The common `FormTemplate` is checking if the function exists and if it does, it will render the `cancel` button. As `useCallback` always returns a function regardless of the callback, it confuses the templates.

Comment (zgover) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js:56]: @Hyperkid123 I'll leave that to you to propose a solution. It appears there are a few areas that need addressing with the test cancel validation. Personally, I think this should only be validating the `onCancel` pre and post logic, if currently the default behavior displays a cancel element, that should instead be decided by the lib consumer.

Comment (Hyperkid123) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js:56]: Yes, they can decide. We are talking about common default templates that are not mandatory to use. So ultimately the decision is on the consumer. I would not like to introduce an entirely new prop that says `enableCancel` or something similar.

You should be able to fix with with a simple `onCancel: onCancel && handleCancelCallback(getState),` assignment

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (zgover) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js:57]: @Hyperkid123 Am I missing something? I thought I copied directly almost the exact thing that was there before. The useCallback initial return is called immediately before passing to the context provider.

https://github.com/data-driven-forms/react-forms/blob/3cfbaf6f88ba81a8aae94be2ffd9bcf2c59c1cc0/packages/react-form-renderer/src/form-renderer/form-renderer.js#L85

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (Hyperkid123) [inline packages/react-form-renderer/src/form-renderer/form-renderer.js:140]: The check should be here:
```jsx
onCancel: onCancel && handleCancelCallback(getState),
```

File: packages/react-form-renderer/src/form-renderer/form-renderer.js
Comment (Hyperkid123) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js:57]: The issue is that the `onCancel` passed to the renderer is always a function. (use call back will always return a function). Thus the cancel buttons show even if the user did not define any `onCancel` function. Even if you return `undefined` from the callback, it will generate a `() => undefined` function which is still going to trigger the conidtion.

Comment (Hyperkid123) [reply packages/react-form-renderer/src/form-renderer/form-renderer.js:57]: The implementation of the `handleCallback` was OK before.

**Comment by zgover** [comment]
Seriously, why are these lint checks setup as errors??? No wonder there aren't very many contributions to this repo. It is absolutely far too aggressive, and if that's what's preferred cool, but keep it **preferred** i.e., warn, but not error required, especially for such minor changes. They're not blocking anything besides the build. All of the lint errors caused in this contribution has made my entire experience of reveling my efforts for this lib utterly annoying and does not inspire me to make future updates.

I'd understand it the lint check was on line-endings, multiple empty lines, semicolons, line max column etc... 

But not single line objects, spaces in open/close curly brackets but not square brackets, forced line breaks for lambda/arrow, or forced single line lambda/arrow, disallowing parenthesis to wrap returns/conditional JSX, or shorthands of lambda/arrow functions, array assignments, JSX attribute props, etc.. Merging it all into a single line as required for many of these items limit readability for future maintenance, consumption and overall cleanliness. But if they're left to warnings then reviewers can verify if the warns are ok to approve/merge or if they're outstanding and the code written is ugly/poorly organized then require the lint checks by request.

**Comment by Hyperkid123** [comment]
> They're not blocking anything besides the build. All of the lint errors caused in this contribution has made my entire experience of reveling my efforts for this lib utterly annoying and does not inspire me to make future updates.

You can use the automatic eslint fix command and it will do all the formatting for you. 
```sh
yarn lint --fix
```
(or embedded IDE command)

It took me about 10 seconds to fix all but 3 linting errors and they were related to prop types warnings.

**Comment by rvsia** [comment]
@zgover Do you use an editor or an IDE with ESLINT plugin? To solve all lint errors takes only pressing one keyboard shortcut. Or you can run `yarn lint —fix` command to automatically fix issues. Thanks to prettier, almost all errors can be fixed automatically and we have never heard feedback like this from anyone else yet. And it definitely helps to keep the code cleaner.

**Comment by zgover** [comment]
> @zgover Do you use an editor or an IDE with ESLINT plugin? To solve all lint errors takes only pressing one keyboard shortcut. Or you can run `yarn lint —fix` command to automatically fix issues. Thanks to prettier, almost all errors can be fixed automatically and we have never heard feedback like this from anyone else yet. And it definitely helps to keep the code cleaner.

@rvsia yes I use IntelliJ IDEA for large projects and VSCode for adhoc file viewing. But for very small changes such as this began when starting, I typically make edits in the GitHub web app. Being that it typically takes me more time to pull down a local copy of the branch setup the IDE working space in either one of the editors—all in the amount of time I could make the edits, send an update or PR and be done.

**Comment by zgover** [comment]
> > They're not blocking anything besides the build. All of the lint errors caused in this contribution has made my entire experience of reveling my efforts for this lib utterly annoying and does not inspire me to make future updates.
> 
> 
> 
> You can use the automatic eslint fix command and it will do all the formatting for you. 
> 
> ```sh
> 
> yarn lint --fix
> 
> ```
> 
> (or embedded IDE command)
> 
> 
> 
> It took me about 10 seconds to fix all but 3 linting errors and they were related to prop types warnings.

Lol my point still stands though for such small linting issues they should be warnings, it makes it quick and fluid to make edits in the web app. It's something worth considering being that it's not as though what was written was a mess but rather easier on the eyes for readability.

**Comment by zgover** [comment]
Y'all are more than welcome to partake and not just advise lol, I'm almost positive the option to allow reviewers to edit the PR branch. Particularly, I am not too familiar with the tests in the repo and it would be highly more efficient for y'all to write the last few parts i.e., tests and docs.

**Comment by Hyperkid123** [comment]
> I am not too familiar with the tests in the repo and it would be highly more efficient for y'all to write the last few parts i.e., tests and docs.

We will do the rest of the busy work.

**Comment by Hyperkid123** [comment]
@zgover I had to do one small change to the children as a ReactNode variant to propagate the props. You can check this commit: [fix(renderer): propagate template props to children node](https://github.com/data-driven-forms/react-forms/pull/1218/commits/a6a30b52f579770d51289387aaa2fa7080af69e9)

File: packages/react-renderer-demo/src/pages/components/children.md
Comment (rvsia) [inline packages/react-renderer-demo/src/pages/components/children.md:28]: empty line

File: packages/react-renderer-demo/src/pages/components/renderer.md
Comment (rvsia) [inline packages/react-renderer-demo/src/pages/components/renderer.md:52]: Can we put a link to children here? 

[Read more](/components/form-template).

**Review by rvsia — COMMENTED** [review]
Just some small changes to the documentation.

@kp0zhiqian would it be possible to check (even later) changes in the documentation? Thanks! 🙇

File: packages/react-renderer-demo/src/pages/components/children.md
Comment (kp0zhiqian) [reply packages/react-renderer-demo/src/pages/components/children.md:7]: emm, I'm not sure what you need me to do? or maybe @ the wrong person?🧐

Comment (rvsia) [reply packages/react-renderer-demo/src/pages/components/children.md:7]: Yep, I did, I am sorry 😅 

@yiranc0205 Please could you take a look? 🙏

**Comment by yiranc0205** [comment]
No problem, I will check it later😊

Richard Všianský ***@***.***> 于2022年4月7日周四 16:21写道：

**Review by rvsia — APPROVED** [review]

**Review by zgover — COMMENTED** [review]
@Hyperkid123 was this ever addressed? I navigated away before submitting the changes review with my concerns inside.

File: packages/react-renderer-demo/src/pages/components/children.md
Comment (yiranc0205) [inline packages/react-renderer-demo/src/pages/components/children.md:58]: "example" in the end should be "examples".

File: packages/react-renderer-demo/src/pages/components/renderer.md
Comment (yiranc0205) [inline packages/react-renderer-demo/src/pages/components/renderer.md:40]: "defines" should be "define"

**Pull request merged**
